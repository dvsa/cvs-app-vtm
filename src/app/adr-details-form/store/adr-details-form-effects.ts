import {Injectable} from '@angular/core';
import {Actions, Effect, EffectNotification, ofType, OnRunEffects} from '@ngrx/effects';
import {Observable} from 'rxjs';
import {tap, exhaustMap} from 'rxjs/operators';
import {LoadAction} from '@app/adr-details-form/store/adrDetails.actions';


@Injectable()
export class AdrDetailsFormEffects implements OnRunEffects {
  @Effect({ dispatch: false })
  load$ = this.actions$.pipe(
    ofType<LoadAction>(LoadAction.TYPE),
    tap(action => {
      console.log('In AdrDetailsFormEffects payload =>', action.payload);
    })
  );

  ngrxOnRunEffects(resolvedEffects$: Observable<EffectNotification>) {
    return this.actions$.pipe(
      exhaustMap((res) => {
        return resolvedEffects$;
      })
    );
  }

  constructor(
    private actions$: Actions,
  ) { }
}
