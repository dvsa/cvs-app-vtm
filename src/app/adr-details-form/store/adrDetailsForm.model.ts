import { Boxed } from "ngrx-forms";

export interface approvalDate {
  day: number;
  month: number;
  year: number;
}

export interface adrDetailsFormModel {
  name: string;
  street: string;
  town: string;
  city: string;
  postcode: string;
  type: string;
  approvalDate: approvalDate;
  permittedDangerousGoods: { [id: string]: boolean };
  compatibilityJ: boolean;
  additionalNotes: { [id: string]: boolean };
  adrTypeApprovalNo: string;
  tankManufacturer: string;
  yearOfManufacture: number;
  tankManufacturerSerialNo: string;
  tankTypeAppNo: string;
  tankCode: string;
  tankDocuments: (string | ArrayBuffer)[];
  substancesPermitted: string;
  selectReferenceNumber: string;
  statement: string;
  productListRefNo: string;
  productListUnNo: number[];
  productList: string;
  specialProvisions: string;
  tc2Type: string;
  tc2IntermediateApprovalNo: string;
  tc2IntermediateExpiryDate: {
    day: number;
    month: number;
    year: number;
  };
  tc3Type: string;
  tc3PeriodicNumber: string;
  tc3PeriodicExpiryDate: {
    day: number;
    month: number;
    year: number;
  }[];
  memosApply: { [id: string]: boolean };
  listStatementApplicable: '';
  batteryListNumber: string;
  brakeDeclarationIssuer: string;
  brakeEndurance: string;
  brakeDeclarationsSeen: string;
  declarationsSeen: string;
  weight: number;
  certificateReq: string;
  adrMoreDetail: string;
}

