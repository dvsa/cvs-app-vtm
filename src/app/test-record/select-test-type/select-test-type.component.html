<p *ngIf="testRecordObservable$ | async  as testRecord">
  {{ testRecord.testType | json }}
</p>
