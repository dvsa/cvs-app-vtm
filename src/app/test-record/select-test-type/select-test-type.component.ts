import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  EventEmitter,
  Output
} from '@angular/core';
import { KeyValue } from '@angular/common';
import { TreeData } from '@app/models/tree-data';

@Component({
  selector: 'vtm-select-test-type',
  templateUrl: './select-test-type.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class SelectTestTypeComponent implements OnInit {
  @Input() filteredCategories: TreeData[];
  @Output() testTypeSelected = new EventEmitter<KeyValue<string, string>>();
  testTypeData: KeyValue<string, string>;

  constructor() {}

  ngOnInit() {}

  selectedTestTypeHandler(testTypeData: KeyValue<string, string>) {
    this.testTypeData = testTypeData;
  }

  updateSelectedTestResult() {
    this.testTypeSelected.emit(this.testTypeData);
  }
}
