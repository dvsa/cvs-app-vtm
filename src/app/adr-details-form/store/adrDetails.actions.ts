import { Action } from '@ngrx/store';
import { AddArrayControlAction, RemoveArrayControlAction } from 'ngrx-forms';
import { adrDetailsFormModel } from './adrDetailsForm.model';

export class LoadAction implements Action {
  static readonly TYPE = 'adrDetails/LOAD';
  readonly type = LoadAction.TYPE;
  constructor(public payload: any) { }
}

export class SetSubmittedValueAction implements Action {
  static readonly TYPE = 'adrDetails/SET_SUBMITTED_VALUE';
  readonly type = SetSubmittedValueAction.TYPE;
  constructor(public submittedValue: adrDetailsFormModel) { }
}

export class CreatePermittedDangerousGoodElementAction implements Action {
  static readonly TYPE = 'adrDetails/CREATE_PERMITTED_DANGEROUS_GOOD_ELEMENT';
  readonly type = CreatePermittedDangerousGoodElementAction.TYPE;
  constructor(public name: string, public payload: boolean = false) { }
}

export class RemovePermittedDangerousGoodElementAction implements Action {
  static readonly TYPE = 'adrDetails/REMOVE_PERMITTED_DANGEROUS_GOOD_ELEMENT';
  readonly type = RemovePermittedDangerousGoodElementAction.TYPE;
  constructor(public name: string) { }
}

export class CreateGuidanceNoteElementAction implements Action {
  static readonly TYPE = 'adrDetails/CREATE_GUIDANCE_NOTE_ELEMENT';
  readonly type = CreateGuidanceNoteElementAction.TYPE;
  constructor(public name: string, public payload: boolean = false) { }
}

export class RemoveGuidanceNoteElementAction implements Action {
  static readonly TYPE = 'adrDetails/REMOVE_GUIDANCE_NOTE_ELEMENT';
  readonly type = RemoveGuidanceNoteElementAction.TYPE;
  constructor(public name: string) { }
}

export class CreateProductListUnNoElementAction implements Action {
  static readonly TYPE = 'adrDetails/CREATE_PRODUCTLIST_UNNO_ELEMENT';
  readonly type = CreateProductListUnNoElementAction.TYPE;
  constructor(public name: string, public payload: string) { }
}

export class RemoveProductListUnNoElementAction implements Action {
  static readonly TYPE = 'adrDetails/REMOVE_PRODUCTLIST_UNNO_ELEMENT';
  readonly type = RemoveProductListUnNoElementAction.TYPE;
  constructor(public name: string) { }
}

export class CreateTc3TypeElementAction extends AddArrayControlAction<number>  {
  static readonly SUB_TYPE = 'adrDetails/CREATE_TC3_TYPE_ELEMENT';
  public readonly subtype = CreateTc3TypeElementAction.SUB_TYPE;
  constructor(public controlId: string, public value: any, index?: number) {
    super(controlId, value, index);
  }
}

export class CreateTc3PeriodicNumberElementAction extends AddArrayControlAction<number> {
  static readonly SUB_TYPE = 'adrDetails/CREATE_TC3_PERIODIC_NUMBER_ELEMENT';
  readonly subtype = CreateTc3PeriodicNumberElementAction.SUB_TYPE;
  constructor(public controlId: string, public value: any, index?: number) {
    super(controlId, value, index);
  }
}

export class CreateTc3PeriodicExpiryDateElementAction extends AddArrayControlAction<{ day: number; month: number; year: number; }> {
  static readonly SUB_TYPE = 'adrDetails/CREATE_TC3_PERIODIC_EXPIRY_DATE_ELEMENT';
  readonly subtype = CreateTc3PeriodicExpiryDateElementAction.SUB_TYPE;
  constructor(public controlId: string, public value: any, index?: number) {
    super(controlId, value, index);
  }
}


export class DownloadDocumentFileAction implements Action {
  static readonly TYPE = 'technicalRecords/DOWNLOAD_DOCUMENT_FILE';
  readonly type = DownloadDocumentFileAction.TYPE;
  constructor(public filename: string) {
  }
}

export class DownloadDocumentFileActionSuccess implements Action {
  static readonly TYPE = 'technicalRecords/DOWNLOAD_DOCUMENT_FILE_SUCCESS';
  readonly type = DownloadDocumentFileActionSuccess.TYPE;
  constructor(public payload: { blob: Blob, fileName?: string }) {
  }
}

export class DownloadDocumentFileActionFailure implements Action {
  static readonly TYPE = 'technicalRecords/DOWNLOAD_DOCUMENT_FILE_FAILURE';
  readonly type = DownloadDocumentFileActionFailure.TYPE;
  constructor(public payload: any) {
  }
}

export class AddTankDocumentAction extends AddArrayControlAction<(string | ArrayBuffer)> {
  static readonly SUB_TYPE = 'adrDetails/ADD_TANK_DOCUMENT';
  readonly subtype = AddTankDocumentAction.SUB_TYPE;
  constructor(public controlId: string, public value: string | ArrayBuffer, index?: number) {
    super(controlId, value, index);
  }
}

export class RemoveTankDocumentAction extends RemoveArrayControlAction {
  static readonly SUB_TYPE = 'adrDetails/REMOVE_TANK_DOCUMENT';
  readonly subtype = RemoveTankDocumentAction.SUB_TYPE;
  constructor(public controlId: string, index?: number) {
    super(controlId, index);
  }
}

