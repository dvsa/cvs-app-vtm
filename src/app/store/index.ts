import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { environment } from '../../environments/environment';
import { initialState as initialTechnicalRecordsState, STORE_FEATURE_TECHNICAL_RECORDS_KEY, TechnicalRecordServiceState, vehicleTechRecordReducer } from './technical-records/technical-record-service.reducer';
import { initialTestResultsState, STORE_FEATURE_TEST_RESULTS_KEY, testResultsReducer, TestResultsState } from './test-records';
import { initialState as initialUserState, STORE_FEATURE_USER_KEY, userServiceReducer, UserServiceState } from './user/user-service.reducer';
import { globalErrorReducer, GlobalErrorState, initialGlobalErrorState, STORE_GLOBAL_ERROR_KEY } from '@store/global-error/reducers/global-error-service.reducer';

export interface State {
  [STORE_FEATURE_USER_KEY]: UserServiceState;
  [STORE_FEATURE_TECHNICAL_RECORDS_KEY]: TechnicalRecordServiceState;
  [STORE_FEATURE_TEST_RESULTS_KEY]: TestResultsState;
  [STORE_GLOBAL_ERROR_KEY]: GlobalErrorState;
  router?: any;
}

export const initialAppState = {
  [STORE_FEATURE_USER_KEY]: initialUserState,
  [STORE_FEATURE_TECHNICAL_RECORDS_KEY]: initialTechnicalRecordsState,
  [STORE_FEATURE_TEST_RESULTS_KEY]: initialTestResultsState,
  [STORE_GLOBAL_ERROR_KEY]: initialGlobalErrorState
};

export const reducers: ActionReducerMap<State> = {
  [STORE_FEATURE_USER_KEY]: userServiceReducer,
  [STORE_FEATURE_TECHNICAL_RECORDS_KEY]: vehicleTechRecordReducer,
  [STORE_FEATURE_TEST_RESULTS_KEY]: testResultsReducer,
  [STORE_GLOBAL_ERROR_KEY]: globalErrorReducer
};

export const metaReducers: MetaReducer<State>[] = !environment.production ? [] : [];
