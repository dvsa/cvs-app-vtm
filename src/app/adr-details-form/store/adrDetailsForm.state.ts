import { createFormGroupState, FormGroupState, box } from 'ngrx-forms';
import { adrDetailsFormModel } from './adrDetailsForm.model';
import {IAppState as RootState} from '../../store/state/app.state';

export interface IAppState extends RootState {
  adrDetails: {
    formState: FormGroupState<adrDetailsFormModel>;
    permittedDangerousGoodsOptions: string[];
    additionalNotesOptions: string[];
    productListUnNo: {
      maxIndex: number;
      options: number[];
    };
    tc3Type: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicNumber: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicExpiryDate: {
      maxIndex: number;
      options: { day: number; month: number; year: number; }[];
    };
    tankDocuments: {
      maxIndex: number;
      options: (string | ArrayBuffer) [];
    };
    submittedValue: adrDetailsFormModel | undefined;
  };
}

export const FORM_ID = 'adrDetails';

export const INITIAL_STATE = createFormGroupState<adrDetailsFormModel>(FORM_ID, {
  name: '',
  street: '',
  town: '',
  city: '',
  postcode: '',
  type: '',
  approvalDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  permittedDangerousGoods: {
  },
  compatibilityJ: false,
  additionalNotes: {},
  adrTypeApprovalNo: '',
  tankManufacturer: '',
  yearOfManufacture: 1,
  tankManufacturerSerialNo: '',
  tankTypeAppNo: '',
  tankCode: '',
  tankDocuments:[],
  substancesPermitted: '',
  selectReferenceNumber: '',
  statement: '',
  productListRefNo: '',
  productListUnNo: [],
  productList: '',
  specialProvisions: '',
  tc2Type: '',
  tc2IntermediateApprovalNo: '',
  tc2IntermediateExpiryDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  tc3Type: [],
  tc3PeriodicNumber: [],
  tc3PeriodicExpiryDate: [],
  memosApply: { '07/09 3mth leak ext': false },
  listStatementApplicable: '',
  batteryListNumber: '',
  brakeDeclarationIssuer: '',
  brakeEndurance: '',
  brakeDeclarationsSeen: '',
  declarationsSeen: '',
  weight: 0,
  certificateReq: '',
  adrMoreDetail: ''
});

export function createInitialState(initialAdrDetails: any): any {
  return {
    name: initialAdrDetails.applicantDetails.name,
    street: initialAdrDetails.applicantDetails.street,
    town: initialAdrDetails.applicantDetails.town,
    city: initialAdrDetails.applicantDetails.city,
    postcode: initialAdrDetails.applicantDetails.postcode,
    type: initialAdrDetails.vehicleDetails.type,
    approvalDate: {
      day: initialAdrDetails.vehicleDetails.approvalDate.split('-')[2],
      month: initialAdrDetails.vehicleDetails.approvalDate.split('-')[1],
      year: initialAdrDetails.vehicleDetails.approvalDate.split('-')[0]
    },
    permittedDangerousGoods: {
    },
    compatibilityJ: false,
    additionalNotes: {},
    adrTypeApprovalNo: '',
    tankManufacturer: initialAdrDetails.tank.tankDetails.tankManufacturer,
    yearOfManufacture: initialAdrDetails.tank.tankDetails.yearOfManufacture,
    tankManufacturerSerialNo: initialAdrDetails.tank.tankDetails.tankManufacturerSerialNo,
    tankTypeAppNo: initialAdrDetails.tank.tankDetails.tankTypeAppNo,
    tankCode: initialAdrDetails.tank.tankDetails.tankCode,
    tankDocuments: initialAdrDetails.documents,
    substancesPermitted: '',
    selectReferenceNumber: '',
    statement: '',
    productListRefNo: '',
    productListUnNo: [],
    productList: '',
    specialProvisions: '',
    tc2Type: '',
    tc2IntermediateApprovalNo: '',
    tc2IntermediateExpiryDate: {
      day: 1,
      month: 1,
      year: 1920
    },
    tc3Type: [],
    tc3PeriodicNumber: [],
    tc3PeriodicExpiryDate: [],
    memosApply: { '07/09 3mth leak ext': false },
    listStatementApplicable: '',
    batteryListNumber: '',
    brakeDeclarationIssuer: '',
    brakeEndurance: '',
    brakeDeclarationsSeen: '',
    declarationsSeen: '',
    weight: 0,
    certificateReq: '',
    adrMoreDetail: ''
  };
}
