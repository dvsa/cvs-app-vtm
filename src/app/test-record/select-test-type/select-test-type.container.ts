import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs';
import { Store } from '@ngrx/store';

import { IAppState } from '@app/store/state/app.state';
import {
  getActiveTechRecord,
  getFilteredTestTypeCategories,
  getTestViewState,
  selectTestType
} from '@app/store/selectors/VehicleTestResult.selectors';
import {
  CreateTestResult,
  UpdateSelectedTestResultModel
} from '@app/store/actions/VehicleTestResult.actions';
import { VIEW_STATE } from '@app/app.enums';
import { TreeData } from '@app/models/tree-data';
import { TechRecord } from '@app/models/tech-record.model';
import { TestRecordTestType } from '@app/models/test-record-test-type';

@Component({
  selector: 'vtm-select-test-type-container',
  template: `
    <ng-container *ngIf="filteredCategories$ | async as filteredCategories">
      <vtm-select-test-type
        [filteredCategories]="filteredCategories"
        [currentState]="viewState$ | async"
        [vehicleRecord]="activeRecord$ | async"
        (testTypeSelected)="updateSelectedTestResult($event)"
      >
      </vtm-select-test-type>
    </ng-container>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class SelectTestTypeContainer implements OnInit {
  testRecord$: Observable<TestRecordTestType>;
  filteredCategories$: Observable<TreeData[]>;
  viewState$: Observable<VIEW_STATE>;
  activeRecord$: Observable<TechRecord>;

  constructor(private store: Store<IAppState>) {}

  ngOnInit(): void {
    this.testRecord$ = this.store.select(selectTestType);
    this.activeRecord$ = this.store.select(getActiveTechRecord);
    this.filteredCategories$ = this.store.select(getFilteredTestTypeCategories);
    this.viewState$ = this.store.select(getTestViewState);
  }

  updateSelectedTestResult(newTestResultData) {
    if (!!newTestResultData['key']) {
      this.store.dispatch(new UpdateSelectedTestResultModel(newTestResultData));
    } else {
      this.store.dispatch(new CreateTestResult(newTestResultData));
    }
  }
}
