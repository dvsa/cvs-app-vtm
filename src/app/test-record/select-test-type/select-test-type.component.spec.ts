import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { SelectTestTypeComponent } from './select-test-type.component';
import { INITIAL_STATE, Store } from '@ngrx/store';
import { hot } from 'jasmine-marbles';
import { ActivatedRoute, convertToParamMap } from '@angular/router';
import { of } from 'rxjs';

describe('SelectTestTypeComponent', () => {
  let component: SelectTestTypeComponent;
  let fixture: ComponentFixture<SelectTestTypeComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [SelectTestTypeComponent],
      providers: [
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => hot('-a', { a: INITIAL_STATE })),
            select: jest.fn()
          }
        },
        {
          provide: ActivatedRoute,
          useValue: {
            queryParams: of({ id: 'W01A34247' }),
            paramMap: of(convertToParamMap({ id: 'W01A34247' }))
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SelectTestTypeComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
