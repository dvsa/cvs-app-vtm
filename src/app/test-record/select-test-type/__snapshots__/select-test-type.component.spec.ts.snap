// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectTestTypeComponent should create 1`] = `
<vtm-select-test-type
  route={[Function Object]}
  store={[Function Object]}
  testRecordObservable$="undefined"
>
  
</vtm-select-test-type>
`;
